model/task_lifecycle.go

```diff
@@ -775,48 +775,104 @@ func RestartItemsAfterVersion(cq *commitqueue.CommitQueue, project, version, cal
+	}
+
+	_, err := dequeueAndRestartItem(dequeueAndRestartOptions{
+		cq:           cq,
+		projectID:    t.Project,
+		versionID:    t.Version,
+		taskID:       t.Id,
+		caller:       caller,
+		reason:       reason,
+		mergeErrMsg:  mergeErrMsg,
+		githubStatus: githubState,
+	})
+	return err
+}
+
+// DequeueAndRestartForVersion restarts all items after the commit queue
+// item, aborts/dequeues this version, and sends an updated status to GitHub. If
+// it succeeds, it returns the removed item.
+func DequeueAndRestartForVersion(cq *commitqueue.CommitQueue, project, version, user, reason string) (*commitqueue.CommitQueueItem, error) {
+	return dequeueAndRestartItem(dequeueAndRestartOptions{
+		cq:           cq,
+		projectID:    project,
+		versionID:    version,
+		caller:       user,
+		reason:       reason,
+		mergeErrMsg:  fmt.Sprintf("commit queue item '%s' is being dequeued: %s", version, reason),
+		githubStatus: message.GithubStateFailure,
+	})
+}
+
+type dequeueAndRestartOptions struct {
+	cq           *commitqueue.CommitQueue
+	projectID    string
+	versionID    string
+	taskID       string
+	caller       string
+	reason       string
+	mergeErrMsg  string
+	githubStatus message.GithubState
+}
+
+func dequeueAndRestartItem(opts dequeueAndRestartOptions) (*commitqueue.CommitQueueItem, error) {
+	if opts.cq == nil {
+		cq, err := commitqueue.FindOneId(opts.projectID)
 		if err != nil {
-			return errors.Wrapf(err, "getting commit queue for project '%s'", t.Project)
+			return nil, errors.Wrapf(err, "getting commit queue for project '%s'", opts.projectID)
 		}
 		if cq == nil {
-			return errors.Errorf("commit queue for project '%s' not found", t.Project)
+			return nil, errors.Errorf("commit queue for project '%s' not found", opts.projectID)
 		}
+		opts.cq = cq
 	}
-	// this must be done before dequeuing so that we know which entries to restart
-	if err := RestartItemsAfterVersion(cq, t.Project, t.Version, caller); err != nil {
-		return errors.Wrapf(err, "restarting items after version '%s'", t.Version)
+
+	// Restart later items before dequeueing this item so that we know which
+	// entries to restart.
+	if err := RestartItemsAfterVersion(opts.cq, opts.projectID, opts.versionID, opts.caller); err != nil {
+		return nil, errors.Wrap(err, "restarting later commit queue items")
 	}
 
-	p, err := patch.FindOneId(t.Version)
+	p, err := patch.FindOneId(opts.versionID)
 	if err != nil {
-		return errors.Wrap(err, "finding patch")
+		return nil, errors.Wrapf(err, "finding patch '%s'", opts.versionID)
 	}
 	if p == nil {
-		return errors.Errorf("patch '%s' not found", t.Version)
+		return nil, errors.Errorf("patch '%s' not found", opts.versionID)
 	}
-	if err := tryDequeueAndAbortCommitQueueVersion(p, *cq, t, caller); err != nil {
-		return err
+
+	removed, err := tryDequeueAndAbortCommitQueueVersion(p, *opts.cq, opts.taskID, opts.mergeErrMsg, opts.caller)
```

Comment:
Since we've been referring to them as items for the most part in the function names can we rename this to `tryDequeueAndAbortCommitQueueItem`?

Reply:
Done.